airflow/models/param.py

```diff
@@ -49,10 +51,28 @@ def __init__(self, default: Any = NOTSET, description: Optional[str] = None, **k
         self.description = description
         self.schema = kwargs.pop('schema') if 'schema' in kwargs else kwargs
 
+        if self.has_value:
+            self._validate(self.value, self.schema)
+
+    @staticmethod
+    def _validate(value, schema):
+        """
+        1. Check that value is json-serializable; if not, warn.  In future release we will require
+        the value to be json-serializable.
+        2. Validate ``value`` against ``schema``
+        """
+        try:
+            json.dumps(value)
+        except Exception:
+            warnings.warn(
+                "The use of non-json-serializable params is deprecated and will be removed in "
+                " a future release",
+                DeprecationWarning,
+            )
```

Comment:
```suggestion
            warnings.warn(
                "The use of non-json-serializable params is deprecated and will be removed in "
                "a future release",
                DeprecationWarning,
            )
```

Also I think we should add `stacklevel` to make the warning show where the user defines the params.

Reply:
I experimented with different values of stacklevel and it seems it's not possible to make it reach the place of definition.  It's too nested it seems.